Enable the use of enterprise edition from `neo4j-shell -path ...`

thobe opened the pull request:
The shell (when used as `neo4j-shell -path ...`) always used a community `GraphDatabaseFactory` for starting the database, this meant that it would be unable to open a database that made use of Property Existence Constraints.

This pull request also adds the ability to tell:
1. which version and edition of Neo4j the shell command belongs to (`neo4j-shell -version`), and
2. which edition of Neo4j is running for the database the shell is connected to (through the `dbinfo` command)

thobe committed:
Allow shell to run an enterprise edition database

Also adds a flag (-version) to neo4j-shell to tell which edition and
version of Neo4j the shell is running.

thobe commented:
This should go some way towards resolving #5904.

lutovich commented on community/jmx/src/main/java/org/neo4j/jmx/impl/KernelBean.java:
```diff
@@ -54,6 +54,25 @@
 
+        // perhaps not the most elegant way to tell which edition we are running...
+        String edition;
+        switch ( kernel.graphDatabase().getClass().getSimpleName() )
```
Is it possible to add `GraphDatabaseAPI#getEdition()` and avoid this class name check?

lutovich commented on community/shell/src/main/java/org/neo4j/shell/StartClient.java:
```diff
@@ -156,8 +181,19 @@ void start( String[] arguments, CtrlCHandler signalHandler )
+        {
+            String version = Version.getKernelVersion();
+            String edition = "Community";
+            if ( factory.getClass() != GraphDatabaseFactory.class )
```
`GraphDatabaseFactory` can also have `String #getEdition()` method to avoid class name check

MishaDemianenko commented on community/jmx/src/main/java/org/neo4j/jmx/impl/KernelBean.java:
```diff
@@ -54,6 +54,25 @@
         this.query = support.createMBeanQuery( instanceId );
 
+        // perhaps not the most elegant way to tell which edition we are running...
+        String edition;
```
we already have editionName in https://github.com/neo4j/neo4j/blob/3.0/community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacadeFactory.java#L95, and according to usages it's properly populated

thobe replied:
Ok. Thanks.

MishaDemianenko commented on community/shell/src/main/java/org/neo4j/shell/impl/SystemOutput.java:
```diff
@@ -35,12 +36,22 @@
+    {
+        try
+        {
+            this.out = new PrintWriter( new OutputStreamWriter( out, "UTF-8" ) );
```
StandardCharsets.UTF_8 should help to avoid `UnsupportedEncodingException` handling

thobe replied:
Yes. Not sure I care about fixing up that aspect of this code though.

thobe commented:
@lutovich perhaps. I'd hate to leak more things through those interfaces though.

thobe commented:
Given that this investigation lead to [a workaround for the original issue](https://github.com/neo4j/neo4j/issues/5904#issuecomment-160972219), perhaps we should just not do this change.

thobe committed:
Ensure that edition is available from shell dbinfo

This adds tests that verify that you can get the edition of the database
using neo4j-shell. It also fixes an issue where simple output on local
databases in shell would not be printed.
